<a href="https://colab.research.google.com/github/AlukweJonesTerah/ALGOS/blob/main/Sensor_Data_AFRICA_Air_Quality_Archive_Kisumu_November2018.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [123]:
import pandas as pd
import plotly.express as px

In [94]:
df = pd.read_csv("november_2018_sensor_data_archive.csv")

In [95]:
df.head()

sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value
0  122;SDS011;56;-0.108;34.749;2018-11-24T10:49:0...               
1  122;SDS011;56;-0.108;34.749;2018-11-24T10:49:0...               
2  121;DHT22;56;-0.108;34.749;2018-11-24T10:49:07...               
3  121;DHT22;56;-0.108;34.749;2018-11-24T10:49:07...               
4  122;SDS011;56;-0.108;34.749;2018-11-24T10:52:2...

In [96]:
df.columns

Index(['sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value'], dtype='object')

The structure of the data/dataframe should be

1. sensor_id - 122;
2. sensor_type - SDS011;
3. location - 56;
4. lat - -0.108;
5. lon - 34.749;
6. timestamp - 2018-11-24T10:49:06.216079+00:00;
7. value_type - P2;
8. value - 8.65

I will make a new dataframe with each column aligning to each value as indicated above  

## Problems to face there might be empty rows

In [97]:
df.shape

(66574, 1)

In [98]:
df.isnull().sum()

sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value    0
dtype: int64

In [99]:
new_columns = ["sensor_id", "sensor_type", "location", "lat", "lon", "timestamp", "value_type", "value"]
df[new_columns] = df['sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value'].str.split(";", expand=True)

In [100]:
df.drop("sensor_id;sensor_type;location;lat;lon;timestamp;value_type;value", axis=1, inplace=True)

In [101]:
df.shape

(66574, 8)

In [102]:
df.isnull().sum()

sensor_id      0
sensor_type    0
location       0
lat            0
lon            0
timestamp      0
value_type     0
value          0
dtype: int64

## Task 1 is done
we have a new dataframe of shape (66574, 8) and 0 nulls

next we need to check if all raws are with the correct types.

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66574 entries, 0 to 66573
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sensor_id    66574 non-null  object
 1   sensor_type  66574 non-null  object
 2   location     66574 non-null  object
 3   lat          66574 non-null  object
 4   lon          66574 non-null  object
 5   timestamp    66574 non-null  object
 6   value_type   66574 non-null  object
 7   value        66574 non-null  object
dtypes: object(8)
memory usage: 4.1+ MB


In [104]:
df.head()

sensor_id sensor_type location     lat     lon  \
0       122      SDS011       56  -0.108  34.749   
1       122      SDS011       56  -0.108  34.749   
2       121       DHT22       56  -0.108  34.749   
3       121       DHT22       56  -0.108  34.749   
4       122      SDS011       56  -0.108  34.749   

                          timestamp   value_type  value  
0  2018-11-24T10:49:06.216079+00:00           P2   8.65  
1  2018-11-24T10:49:06.216079+00:00           P1   9.45  
2  2018-11-24T10:49:07.719863+00:00  temperature  31.70  
3  2018-11-24T10:49:07.719863+00:00     humidity  44.80  
4  2018-11-24T10:52:25.067433+00:00           P2   4.65

In [105]:
print(f' \n Sensor_id unique values: {df["sensor_id"].unique()}')
print(f' \n Senosor_type unique values: {df["sensor_type"].unique()}')
print(f' \n Location unique values: {df["location"].unique()}')
print(f' \n Latitude unique values: {df["lat"].unique()}')
print(f' \n Longitude unique values: {df["lon"].unique()}')
print(f' \n Timestamp unique values: {df["timestamp"].unique()}')
print(f' \n Value types unique values: {df["value_type"].unique()}')
print(f' \n Value unique values: {df["value"].unique()}')

 
 Sensor_id unique values: ['122' '121' '84' '83']
 
 Senosor_type unique values: ['SDS011' 'DHT22']
 
 Location unique values: ['56']
 
 Latitude unique values: ['-0.108']
 
 Longitude unique values: ['34.749']
 
 Timestamp unique values: ['2018-11-24T10:49:06.216079+00:00' '2018-11-24T10:49:07.719863+00:00'
 '2018-11-24T10:52:25.067433+00:00' ... '2018-11-30T23:58:58.909584+00:00'
 '2018-11-30T23:59:34.353157+00:00' '2018-11-30T23:59:35.813678+00:00']
 
 Value types unique values: ['P2' 'P1' 'temperature' 'humidity']
 
 Value unique values: ['8.65' '9.45' '31.70' ... '24.43' '60.75' '53.45']


In [106]:
df = df.set_index("timestamp")

In [107]:
df.duplicated().sum()

np.int64(58239)

In [108]:
df.to_csv("Sensor-Data-AFRICA-Air-Quality-Archive-Kisumu-November2018_Cleaded.csv")

## Time Series Prediction Using leaner regression, AR(AutoRegression) and ARMA (AutoRegression Moving Average) models

In [113]:
df_= pd.read_csv("Sensor-Data-AFRICA-Air-Quality-Archive-Kisumu-November2018_Cleaded.csv")
df.head()

sensor_id sensor_type location     lat  \
timestamp                                                                 
2018-11-24 10:49:06.216079+00:00       122      SDS011       56  -0.108   
2018-11-24 10:49:06.216079+00:00       122      SDS011       56  -0.108   
2018-11-24 10:49:07.719863+00:00       121       DHT22       56  -0.108   
2018-11-24 10:49:07.719863+00:00       121       DHT22       56  -0.108   
2018-11-24 10:52:25.067433+00:00       122      SDS011       56  -0.108   

                                     lon   value_type  value  
timestamp                                                     
2018-11-24 10:49:06.216079+00:00  34.749           P2   8.65  
2018-11-24 10:49:06.216079+00:00  34.749           P1   9.45  
2018-11-24 10:49:07.719863+00:00  34.749  temperature  31.70  
2018-11-24 10:49:07.719863+00:00  34.749     humidity  44.80  
2018-11-24 10:52:25.067433+00:00  34.749           P2   4.65

In [117]:
#  coverting the selected columns to numerical data values and datetime timestamp index to date

df_[["sensor_id", "location", "lat", "lon", "value"]] = df_[["sensor_id", "location", "lat", "lon", "value"]].apply(pd.to_numeric)
df_.index = pd.DatetimeIndex(df_.index)

In [118]:
df_.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 66574 entries, 1970-01-01 00:00:00 to 1970-01-01 00:00:00.000066573
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   timestamp    66574 non-null  object 
 1   sensor_id    66574 non-null  int64  
 2   sensor_type  66574 non-null  object 
 3   location     66574 non-null  int64  
 4   lat          66574 non-null  float64
 5   lon          66574 non-null  float64
 6   value_type   66574 non-null  object 
 7   value        66574 non-null  float64
dtypes: float64(3), int64(2), object(3)
memory usage: 4.6+ MB


In [119]:
df_.index = df_.index.tz_localize("UTC").tz_convert("Africa/Nairobi")

In [121]:
df.head()

sensor_id sensor_type location     lat  \
timestamp                                                                 
2018-11-24 10:49:06.216079+00:00       122      SDS011       56  -0.108   
2018-11-24 10:49:06.216079+00:00       122      SDS011       56  -0.108   
2018-11-24 10:49:07.719863+00:00       121       DHT22       56  -0.108   
2018-11-24 10:49:07.719863+00:00       121       DHT22       56  -0.108   
2018-11-24 10:52:25.067433+00:00       122      SDS011       56  -0.108   

                                     lon   value_type  value  
timestamp                                                     
2018-11-24 10:49:06.216079+00:00  34.749           P2   8.65  
2018-11-24 10:49:06.216079+00:00  34.749           P1   9.45  
2018-11-24 10:49:07.719863+00:00  34.749  temperature  31.70  
2018-11-24 10:49:07.719863+00:00  34.749     humidity  44.80  
2018-11-24 10:52:25.067433+00:00  34.749           P2   4.65

In [141]:
fig = px.scatter_mapbox(
    df_,
    lat="lat",
    lon="lon",
    color="value_type",
    size="value",
    center={"lat": -0.108, "lon": 34.749},
    mapbox_style="stamen-terrain",
    color_continuous_scale=px.colors.cyclical.IceFire,
    hover_data=["sensor_id", "value_type", "value", "sensor_type"],
    hover_name="sensor_id",
    zoom=10,
    height=600,
    width=800
  )
fig.update_layout(mapbox_style="open-street-map")
fig.show()